### DISCUSSION TITLE
Why is '--no-gc-sections' added in build with MAIN_MODULE/SIDE_MODULE=1?

### DISCUSSION THREAD

**Discussion by alexander-veselov** [emscripten-core/emscripten#16926] [Q&A] [answered]
https://github.com/emscripten-core/emscripten/blob/a62c8119ac1fa1ab21f5dd91c9e7dc92a4de1aad/tools/building.py#L298

--no-gc-sections
 	Disable garbage collection of unused sections.
 
 '--gc-sections' greatly decreases function count (which helps us not to run into the limit of functions). But we can't use it because of '--no-gc-sections'. Why was it added? Is there a way to bypass this and will it be correct?

**Comment by sbc100** [comment, marked as answer]
Because with `-sMAIN_MODULE=1` we explicitly want to include everything.    If you want to DCE your main module you need to build with `MAIN_MODULE=2`.   That is the main difference between =1 and =2.   One does DCE and the other conservatively keeps everything alive in case it is needed at runtime.

**Reply by Maksim-Bozhko** [reply]
@sbc100  Is there a setting that does DCE but keeps exports(symbols marked with visibiliy("default") attribute)? Yes I know there is EMSCRIPTEN_KEEPALIVE macro, but it doesn't work on exported classes.

**Reply by sbc100** [reply]
I would imagine that that is how `MAIN_MODULE=1` should work, although I'm not sure it does today.   Is it just the things that `--gc-sections` removes that you are interested in considering dead?  Does removing the `--no-gc-sections` do what you want?    If so we could look into making that the default for MAIN_MODULE=1 and SIDE_MODULE=1.

**Reply by Maksim-Bozhko** [reply]
Yes, I think so, let me double check function counts and compare. I will tell you for sure soon.

**Reply by Maksim-Bozhko** [reply]
So I counted functions in side module and in main module, for 3 cases:
SIDE_MODULE=1 && MAIN_MODULE=1
main : [funcs]: 25049
side: [funcs]: 3157

SIDE_MODULE=2 && MAIN_MODULE=2
main : [funcs] : 10760
side: [funcs] : 1774

SIDE_MODULE=1 && MAIN_MODULE=1 with cmd.append('--gc-sections') 
main : [funcs] : 21451
side: [funcs] : 2183

So =2 seems to be more agressive still, but it is close

**Reply by Maksim-Bozhko** [reply]
@sbc100 so to answer your initial question: Yes we would need an option to turn on --gc-sections. If we remove --no-gc-sections and add --gc-sections isntead, everything seems to work with SIDE_MODULE/MAIN_MODULE=1 and we didn't see cases of removing anything that we actually needed, I assume because it respects exports and assumes all exports are needed. It also reduces size of wasm_apply_data_reloc function dramatically which allows us to be under function size limit(in combination with -O1)

**Reply by sbc100** [reply]
OK sounds like we can maybe just remove `--no-gc-sections` completely!  I'll give it a try.  Thanks for investigating.

**Reply by sbc100** [reply]
This was fixed in #17056
